### ISSUE TITLE
[Bug]: setWindowOpenHandler only fires for initial window creation when attached to a BrowserWindow, but additionally fires for subsequent window.open calls when attached to a BrowserView

### ISSUE THREAD

**Issue by pushkin-** [electron/electron#29880]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Electron Version

13.1.4

### What operating system are you using?

Windows

### Operating System Version

Windows 10

### What arch are you using?

x64

### Last Known Working Electron version

n/a

### Expected Behavior

I expect that every window.open call invokes the `setWindowOpenHandler` handler.

### Actual Behavior

`setWindowOpenHandler` is only invoked for the initial window open, but not for subsequent navigations via `window.open`.

### Additional Information

I'm not sure if this is a bug or enhancement request, but the `new-window` event would fire for every `window.open` call. It would fire for the initial `window.open(url, "a")` and for subsequent navigations (`window.open(url2, "a")`).

The new handler only fires for the initial window creation.

I was relying on the old behavior for a couple things. I have a concept of tabbed windows. If the custom tab feature was enabled in the `window.open` call, I'd cancel the window.open and create a new window with a BrowserView for a tab. The BrowserView would be navigated to the URL. A subsequent `window.open` with the same window name would make me load the URL into a new tab of the window rather than into the `BrowserWindow` itself.

I can intercept `will-navigate`, but that fires for more things than just `window.open` calls.

There are some other cases where I'm interested in detecting when a subsequent `window.open` is called rather than just any navigation.

**Comment by nornagon** [comment]
@pushkin- I think I understand what the bug is here, but just to make sure we're looking at the same thing, would it be possible for you to provide a small test case for this using [Electron Fiddle](/electron/fiddle)? Thanks 🙇

**Comment by pushkin-** [comment]
[Here's](https://gist.github.com/pushkin-/4e04c535ec255a3dcaab1848a963de55) a gist.
1. Start the app
2. In the devtools, open a window (window.open("https://google.com", "a")
3. Observe the dialog popups up saying the handler was hit
4. Now navigate that window to a page by calling window.open with the same window name
5. You don't get a dialog

[Here's](https://gist.github.com/1d442da9171d0586efa501814e68ecdb) a gist of the old behavior that uses `new-window`. The dialog appears for each window.open call with the same window name.

I have also discovered that this appears to only work when the `new-window` event is hooked for the BrowserView. If you listen for the `new-window` event for a BrowserWindow, it only fires once...

---
Actually, now that I try with a BrowserView with setWindowOpenHandler, it seems to work. I guess I had initially tested only with the BrowserWindow.

Though that raises the question, should the behavior be made consistent?

**Comment by pushkin-** [comment]
And for the record, I prefer the current `BrowserView` behavior where the handler fires for each `window.open` call even if it's only triggering a navigation instead of a new window. Not sure what is correct here, but I would like at least some way of detecting these subsequent `window.open` calls.

**Comment by nornagon** [comment]
Here's a slightly modified version that automatically calls `window.open` rather than relying on devtools:

```js
const {app, BrowserWindow, dialog} = require('electron')

async function createWindow () {
  const mainWindow = new BrowserWindow({
    webPreferences: {
      nativeWindowOpen: true
    }
  })

  await mainWindow.loadURL('https://example.com')

  mainWindow.webContents.setWindowOpenHandler(() => {
    dialog.showMessageBox({message: "setWindowOpenHandler triggered"});
    return { action: "allow" }
  });

  await mainWindow.webContents.executeJavaScript(`window.open("https://google.com", "a"); null`)
  setTimeout(() => {
    mainWindow.webContents.executeJavaScript(`window.open("https://electronjs.org", "a"); null`)
  }, 1000)
}

app.whenReady().then(createWindow)

app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') {
    app.quit()
  }
})

app.on('activate', function () {
  if (BrowserWindow.getAllWindows().length === 0) {
    createWindow()
  }
})
```

What you're saying is that the dialog gets triggered once, but you're expecting it would be triggered twice, yes?

I think this is because Chromium doesn't treat the 2nd `window.open` as a window-open, but rather as a navigation. I'm not sure that invoking the window-open handler in this situation is correct.

**Comment by pushkin-** [comment]
@nornagon Yes, that's correct. But I guess I don't get why it works when using a BrowserView.

**Comment by pushkin-** [comment]
And I wonder if I can rely on that behavior. Because it's quite useful for my case to know when someone calls window.open again to do a navigation.

**Comment by nornagon** [comment]
> it works when using a BrowserView

Any chance you could throw together an example of this "working" behavior? I'm not sure exactly how you're involving browserviews in such a way that this would produce different behavior.

**Comment by pushkin-** [comment]
@nornagon I don't really know what to make of this. In my app, the handler fires for each window.open invocation, though just an hour ago, it didn't for some reason. I also can't repro in a minimal example.

main.js:
```
const {app, BrowserWindow, BrowserView, dialog} = require('electron')

async function createWindow () {
  const mainWindow = new BrowserWindow()
  const view = new BrowserView({ webPreferences: { nativeWindowOpen: true}})
  view.setBounds({x: 0, y:0, width: 600, height: 800})
  mainWindow.setBrowserView(view);

  await view.webContents.loadFile('index.html')
  view.webContents.openDevTools();

  view.webContents.setWindowOpenHandler(() => {
    dialog.showMessageBox({message: "setWindowOpenHandler triggered"});
    return { action: "allow" }
  });
}

app.whenReady().then(createWindow)

app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') {
    app.quit()
  }
})

app.on('activate', function () {
  if (BrowserWindow.getAllWindows().length === 0) {
    createWindow()
  }
})
```

index.html:
```
<button onclick="window.open('https://google.com', 'a')">Open Window</button>
```

I'm going to keep poking around. Not sure what's going on. In Electron 11, `new-window` definitely fired for each window.open call even for navigations.

**Comment by nornagon** [comment]
> In Electron 11, `new-window` definitely fired for each window.open call even for navigations.

even when `nativeWindowOpen: true`?

**Comment by pushkin-** [comment]
Yes, I've only tested with `nativeWindowOpen: true`. However, I'm having trouble reproing with a minimal example, so maybe there's another `webPreferences` setting or something that's causing this behavior. Still poking around.

**Comment by pushkin-** [comment]
I found out what it is. The dialog (new window handler) gets triggered multiple times if the window is opened with `window.opener` disabled (so pass the `noopener` window feature into the window.open call).

Not sure if that's expected, but I'm currently relying on that behavior.

**Comment by pushkin-** [comment]
bump
